hls4ml/model/graph.py

```diff
@@ -604,14 +604,20 @@ def compile(self):
         self._top_function_lib = ctypes.cdll.LoadLibrary(lib_name)
 
     def _get_top_function(self, x):
+
+        io_type = self.config.get_config_value('IOType')
+        interface = self.config.get_config_value('AcceleratorConfig')['Interface'] if self.config.get_config_value('AcceleratorConfig') else None
+        config_weights = (io_type == 'io_stream') and (interface == 'axi_master')
```

Comment:
Two things:
- I don't think IOType of io_stream and Interface of axi_master should be the trigger for reprogrammable weights, I think it would be better as a new configuration parameter (of the AcceleratorConfig). Then there would probably need to be some `assert` somewhere to allow only combinations of those other parameters that have been implemented for (ie right now it also depends on the board)
- As written this can't go here in `ModelGraph` as it's backend specific

Reply:
I agree. This was mostly a placeholder, and I was going to ask what should be better.

As in the main thread, do we already have a matrix or list of the existing possible combinations (what goes with what)? I am not sure how to properly set up the configuration/trigger. Do you have suggestions?

Reply:
I started putting together a [document](https://fermicloud-my.sharepoint.com/:w:/g/personal/gdg_services_fnal_gov/ESO16EampBpBm27vlWeX2jwB8WNf1V25lRKyQOIXkY2EDQ?e=6fMzqV), which is far from being final. @thesps and @jmitrevs should be able to edit. Not sure if that may help or if it is better to keep it to the comments here. I bit of both I guess.

If [I understand it right](https://github.com/fastmachinelearning/hls4ml/blob/main/docs/api/configuration.rst) the additional configuration parameters (to enable programmable weights) should be passed via the function [create_initial_config](https://github.com/fastmachinelearning/hls4ml/blob/563c84c1ddfe5d8298a38ce2cdd74881ac8b955d/hls4ml/backends/vivado_accelerator/vivado_accelerator_backend.py#L62) for the `VivadoAccelerator` backend.

If you agree, then I would add:
- `configurable_weights`
    - `bool`, optional
    - weights are configurable at run time and thus they are exposed to the wrapper interface 
    - defaults to `False`
- `weight_type`
    - `dict`, optional 
    - a dictionary that specifies the data type for set of layer weights (can be `float` or an `ap_type`); if the type is not specified for a layer, then it defaults to `float`
    - defaults to an empty dictionary that is all of the weights are `float`
    - note: VivadoAcceleratorBackend will round the number of bits used to the next power-of-2 value. 

Please let me know if that makes sense.